File: backend/src/AssetBridge.Application/Common/Models/ApiResponse.cs
namespace AssetBridge.Application.Common.Models;

// Standardizes the API response envelope for all endpoints across the system.
// React, Flutter, and external AI clients receive a predictable structure:
// success flag, user-friendly message, payload data, timestamp, and optional errors.
public class ApiResponse<T>
{
    public bool Success { get; set; }
    public string Message { get; set; } = string.Empty;
    public T? Data { get; set; }
    public IDictionary<string, string[]>? Errors { get; set; }
    public DateTime TimestampUtc { get; set; } = DateTime.UtcNow;

    public static ApiResponse<T> SuccessResult(T data, string message = "Operation completed successfully.")
    {
        return new ApiResponse<T>
        {
            Success = true,
            Message = message,
            Data = data,
            TimestampUtc = DateTime.UtcNow
        };
    }

    public static ApiResponse<T> FailureResult(string message, IDictionary<string, string[]>? errors = null)
    {
        return new ApiResponse<T>
        {
            Success = false,
            Message = message,
            Data = default,
            Errors = errors,
            TimestampUtc = DateTime.UtcNow
        };
    }
}

File: mobile/lib/shared/models/api_response.dart
class ApiResponse<T> {
  final bool success;
  final String message;
  final T? data;
  final Map<String, dynamic>? errors;
  final String? timestampUtc;

  ApiResponse({
    required this.success,
    required this.message,
    this.data,
    this.errors,
    this.timestampUtc,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic json) fromJsonT,
  ) {
    return ApiResponse<T>(
      success: json['success'] ?? false,
      message: json['message'] ?? '',
      data: json['data'] != null ? fromJsonT(json['data']) : null,
      errors: json['errors'] as Map<String, dynamic>?,
      timestampUtc: json['timestampUtc'] as String?,
    );
  }
}
